server.js

```diff
@@ -38,168 +43,178 @@ const tracks = {
-					// ...but will have to be this for now...
-					return sendTransaction(OperationsABI, o, 'addChecksum', [`0x000000000000000000000000${commit}`, platform, `0x${sha3}`]);
-				}).then(h => {
-					console.log(`Transaction sent with hash: ${h}`);
-				});
-			}
-		});
+	if (sha3 === '' || !goodTag || !goodPlatform) {
+		throw new Error(`Invalid sha3 (${sha3}), tag (${tag}) or platform (${platform}).`);
 	}
-	res.end(out);
-});
+
+	let body = await fetchFile(commit, '/util/src/misc.rs');
+	const branch = body.match(`const THIS_TRACK. ..static str = "([a-z]*)";`)[1];
+	const track = tracks[branch] ? branch : 'testing';
+
+	console.log(`Track: ${branch} => ${track} (${tracks[track]}) [enabled: ${!!enabledTracks[track]}]`);
+
+	if (!enabledTracks[track]) {
+		throw new Error(`Track not enabled: ${track}`);
+	}
+
+	// make sure the node is running
+	await getNetwork();
+	// Respond already
+	res.send(out);
+
+	const registryAddress = await api.parity.registryAddress();
+	const reg = api.newContract(RegistrarABI, registryAddress);
+	const githubHintAddress = await reg.instance.getAddress.call({}, [githubHint, 'A']);
+
+	console.log(`Registering on GithubHint: ${sha3}, ${url}`);
+	// Should be this...
+	// api.newContract(GitHubHintABI, g).instance.hintURL.postTransaction({from: account.address}, [`0x${sha3}`, url]).then(() => {
+	// ...but will have to be this for now...
+	const hash = await sendTransaction(GitHubHintABI, githubHintAddress, 'hintURL', [`0x${sha3}`, url]);
+	console.log(`Transaction sent with hash: ${hash}`);
+
+	const operationsAddress = reg.instance.getAddress.call({}, [operationsContract, 'A']);
+	console.log(`Registering platform binary: ${commit}, ${platform}, ${sha3}`);
+	// Should be this...
+	// return api.newContract(OperationsABI, o).instance.addChecksum.postTransaction({from: account.address}, [`0x000000000000000000000000${commit}`, platform, `0x${sha3}`]);
+	// ...but will have to be this for now...
+	const hash2 = await sendTransaction(OperationsABI, operationsAddress, 'addChecksum', [`0x000000000000000000000000${commit}`, platform, `0x${sha3}`]);
+	console.log(`Transaction sent with hash: ${hash2}`);
+}));
 
 const server = app.listen(httpPort, function () {
 	const host = server.address().address;
 	const port = server.address().port;
 	console.log('push-release service listening at http://%s:%s', host, port);
 });
 
-const RegistrarABI = require('./res/registrar.json');
-const GitHubHintABI = require('./res/githubhint.json');
-const OperationsABI = require('./res/operations.json');
+function handleAsync (asyncFn) {
+	return (req, res) => asyncFn(req, res)
+		.then(() => {
+			if (!res.headersSent) {
+				throw new Error('No response from handler');
+			}
+		})
+		.catch(err => {
+			console.error(err);
+			res.status(400).end(`Error while processing the request:\n${err.toString()}\n`);
```

Comment:
Should it be a little more fine-grained? Could be a server error at this point.

Reply:
And maybe use `error.message` ?

Reply:
Updated to use `error.message`, I don't care that much about distinguishing between client and server errors. Both indicate some errornous setup, but will consider a change here when I start fixing existing issues.